### PR TITLE
[core] Auto install manager when invalid

### PR DESCRIPTION
**Pull request by yujincheng08**

### CODE REVIEW

File: core/src/main/java/io/github/lsposed/lspd/service/PackageService.java [hunk 1]
```diff
@@ -176,4 +195,87 @@ private static PackageInfo getPackageInfoWithComponents(String packageName, int
...
+            installerConstructor.setAccessible(true);
+            PackageInstaller installer = installerConstructor.newInstance(installerService, installerPackageName, 0);
+            PackageInstaller.SessionParams params = new PackageInstaller.SessionParams(PackageInstaller.SessionParams.MODE_FULL_INSTALL);
+            int installFlags = HiddenApiBridge.PackageInstaller_SessionParams_installFlags(params);
+            installFlags |= 0x00000004/*PackageManager.INSTALL_ALLOW_TEST*/ | 0x00000002/*PackageManager.INSTALL_REPLACE_EXISTING*/;
+            HiddenApiBridge.PackageInstaller_SessionParams_installFlags(params, installFlags);
+
+            int sessionId = installer.createSession(params);
+            try (PackageInstaller.Session session = installer.openSession(sessionId)) {
+                try (InputStream is = new FileInputStream(apkFile); OutputStream os = session.openWrite(apkFile.getName(), 0, -1)) {
```
Comment (aviraxp) [inline core/src/main/java/io/github/lsposed/lspd/service/PackageService.java:257]: seems that is and os is not closed after being used?

Comment (yujincheng08) [reply core/src/main/java/io/github/lsposed/lspd/service/PackageService.java:257]: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

File: core/src/main/java/io/github/lsposed/lspd/service/PackageService.java [hunk 2]
```diff
@@ -235,11 +239,12 @@ public static synchronized boolean installManagerIfAbsent(String packageName, Fi
...
-                return false;
-            // manager is not installed or version not matched, install stub
-            if (pkgInfo != null) {
-                uninstallPackage(new VersionedPackage(pkgInfo.packageName, pkgInfo.versionCode));
+            if (pkgInfo != null && pkgInfo.versionName != null && pkgInfo.applicationInfo != null) {
+                boolean versionMatch = pkgInfo.versionName.equals(BuildConfig.VERSION_NAME);
+                boolean signatureMatch = InstallerVerifier.verifyInstallerSignature(pkgInfo.applicationInfo);
+                if (versionMatch && signatureMatch) return false;
+                if (!signatureMatch || pkgInfo.versionCode > BuildConfig.VERSION_CODE)
+                    uninstallPackage(new VersionedPackage(pkgInfo.packageName, pkgInfo.versionCode));
```
Comment (tehcneko) [inline core/src/main/java/io/github/lsposed/lspd/service/PackageService.java:247]: 这 manager 不就不能单独更新了？

**Review by tehcneko — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/github/lsposed/lspd/service/PackageService.java (hunk 2 above)
Comment (yujincheng08) [reply core/src/main/java/io/github/lsposed/lspd/service/PackageService.java:247]: 能，只要 versionName 一致就可以单独更新

File: core/src/main/java/io/github/lsposed/lspd/service/PackageService.java [hunk 3]
```diff
@@ -176,4 +195,92 @@ private static PackageInfo getPackageInfoWithComponents(String packageName, int
...
+        pm.getPackageInstaller().uninstallExistingPackage(versionedPackage, "com.android.shell", new IntentSenderAdaptor() {
+            @Override
+            public void send(Intent intent) {
+                int status = intent.getIntExtra(PackageInstaller.EXTRA_STATUS, PackageInstaller.STATUS_FAILURE);
+                result[0] = status == PackageInstaller.STATUS_SUCCESS;
+                latch.countDown();
+            }
+        }.getIntentSender(), 0);
+        latch.await();
+        return result[0];
```
Comment (aviraxp) [inline core/src/main/java/io/github/lsposed/lspd/service/PackageService.java:233]: what's the point of returning a boolean if the return value is not used?

Comment (yujincheng08) [reply core/src/main/java/io/github/lsposed/lspd/service/PackageService.java:233]: It's for the manager.

**Review by tehcneko — APPROVED** [review]

**Pull request merged**